### ISSUE TITLE
Adding a colorbar matching a filled contour

### ISSUE THREAD

**Issue by briochemc** [JuliaPlots/Makie.jl#825]
Here is a MWE to start with

```julia
using GLMakie
using KernelDensity
using Random
Random.seed!(0)
# Fake data
N = 1000
x = 2randn(N) .+ 1
y = sin.(x) + randn(N)
# density
D = kde((x, y))
# quantiles of density
δx = step(D.x)
δy = step(D.y)
Q = vec(D.density) * δx * δy
idx = sortperm(Q)
Q_sorted = Q[idx]
Dcum = similar(D.density)
Dcum[idx] .= 1 .- cumsum(Q_sorted)
# plot
fig = Figure()
levels = 0.1:0.1:0.9
ax, co = contourf(fig[1,1], D.x, D.y, Dcum; levels)
fig[1,2] = Colorbar(fig; vertical=true, width=30, levels)
```

which produces something like

<img width="443" alt="Screen Shot 2021-01-25 at 7 21 16 pm" src="https://user-images.githubusercontent.com/4486578/105679408-804de180-5f42-11eb-9a5c-f7204e93dfd7.png">

There are multiple issues here:
1. Colors of the colorbar and filled contour don't match
    - the colors of the filled contour are discrete but those of the colorbar are continuous
    - the levels don't match (the `levels` kwarg is actually not taken into account in the `Colorbar` call)
1. Colors beyond the `levels`: they are not defined here, but it would be great, in the same vein as https://github.com/JuliaPlots/Plots.jl/issues/2266, if we could have an `extend` keyword to add those.

Potentially related Makie.jl issues that I found are https://github.com/JuliaPlots/Makie.jl/issues/406 and https://github.com/JuliaPlots/Makie.jl/issues/596, but maybe there are others?

**Comment by jkrumbiegel** [comment]
I just found an old branch where I already added the triangles for the colorbar, I don't remember why I didn't merge it.

With contourf, what decides if the end bands are shown as triangles or not in the colorbar? They need to be explicit anyway, as opposed to heatmaps where this is really a clipping operation. So you should always know the endpoints and therefore don't really need triangles (although they might look nicer).

Basically the colorbar needs new internals that can draw any arbitrary number of color bins for even / uneven categorical / continuous color maps.

**Comment by briochemc** [comment]
> With contourf, what decides if the end bands are shown as triangles or not in the colorbar?

IMHO, if they are bands with *finite* bounds, they should appear inside. If a band has an infinite bound, I think it should be represented with a triangle. Anyway, as you said, I think the user should specify them. I do like matplotlib's approach, where there is a kwarg `extend` which defaults to `:neither` IIRC. 

<img src="https://camo.githubusercontent.com/4135c297169fac8eb4e06b52969b7b51a7661b51c50eb9f922158471adb83a4a/68747470733a2f2f6d6174706c6f746c69622e6f72672f322e302e302f6d706c5f6578616d706c65732f70796c61625f6578616d706c65732f636f6e746f7572665f64656d6f5f30322e706e67"> 

I'll also copy paste my comment from that issue because I think it's an important bit
> Just to be clear, I am not only asking for the little triangle extensions on the colorbar (which is really a detail), but importantly to actually fill in the countours that go beyond the levels. This corresponds to the yellow and magenta colors in the top right subplot in the image above. This is important to distinguish Missing (or NaN) data from data that is beyond the range of the colorbar.

**Comment by briochemc** [comment]
Should I close this, then?

**Comment by jkrumbiegel** [comment]
If there's not anything missing, sure!

**Comment by briochemc** [comment]
I'll just reopen it if I find a nasty detail to bother you with! 😈 

Joking aside, thanks for the amazing turnover!

**Comment by natgeo-wong** [comment]
Hi, as per advice a few days ago, I would like to reopen this issue for colorbar extensions.

Currently, the :auto: option for extending the colorbar results in the following:
<img width="366" alt="Screen Shot 2022-03-21 at 5 13 02 pm" src="https://user-images.githubusercontent.com/36907442/159806122-d292d616-f49e-432e-b1d5-5d924b6c6075.png">

But I would like something like this instead:
![01c-flsm_BRN](https://user-images.githubusercontent.com/36907442/159806184-00904ac3-cbbf-46a4-9eeb-2f088bc88852.png)

Where the colormap is automatically extended to fit.
